source/extensions/tracers/zipkin/span_buffer.cc

```diff
@@ -16,22 +33,202 @@ bool SpanBuffer::addSpan(const Span& span) {
   return true;
 }
 
-std::string SpanBuffer::toStringifiedJsonArray() {
+SerializerPtr SpanBuffer::makeSerializer(
+    const envoy::config::trace::v2::ZipkinConfig::CollectorEndpointVersion& version,
+    const bool shared_span_context) {
+  switch (version) {
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1:
+    return std::make_unique<JsonV1Serializer>();
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_JSON:
+    return std::make_unique<JsonV2Serializer>(shared_span_context);
+  case envoy::config::trace::v2::ZipkinConfig::HTTP_PROTO:
+    return std::make_unique<ProtobufSerializer>(shared_span_context);
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+std::string JsonV1Serializer::serialize(std::vector<Span>&& zipkin_spans) {
   std::string stringified_json_array = "[";
 
-  if (pendingSpans()) {
-    stringified_json_array += span_buffer_[0].toJson();
-    const uint64_t size = span_buffer_.size();
-    for (uint64_t i = 1; i < size; i++) {
-      stringified_json_array += ",";
-      stringified_json_array += span_buffer_[i].toJson();
+  if (!zipkin_spans.empty()) {
+    absl::StrAppend(&stringified_json_array, zipkin_spans[0].toJson());
+    for (uint64_t i = 1; i < zipkin_spans.size(); i++) {
+      absl::StrAppend(&stringified_json_array, ",", zipkin_spans[i].toJson());
     }
   }
-  stringified_json_array += "]";
+  absl::StrAppend(&stringified_json_array, "]");
 
   return stringified_json_array;
 }
 
+JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
+    : shared_span_context_{shared_span_context} {}
+
+std::string JsonV2Serializer::serialize(std::vector<Span>&& zipkin_spans) {
+  std::vector<zipkin::jsonv2::Span> spans;
+  spans.reserve(zipkin_spans.size() * 2);
```

Comment:
Could we add a comment on why this `*2` is needed?

Reply:
if about list encoding size, something like this is better https://github.com/openzipkin/zipkin/blob/master/zipkin/src/main/java/zipkin2/codec/Encoding.java#L28

Reply:
nevermind me, I don't think it is related to encoding size..

Reply:
This is updated so we don't need to reserve. It was reserving vector size for v2 spans that will be for most twice the size of v1 spans.